### PR TITLE
Continuation of Pr248

### PR DESCRIPTION
**Pull request by adrien-berchet**
Following #248, I made some tests to try to make it work.
It seems ok except for PyPy.

### CODE REVIEW

File: tests/test_functional.py [hunk 1]
```diff
@@ -354,12 +316,14 @@ def setup(self):
...
     def teardown(self):
         session.rollback()
+        session.expunge_all()
         metadata.drop_all()
 
     def _create_one_lake(self):
         lake = Lake(WKTElement('LINESTRING(0 0,1 1)', srid=4326))
         session.add(lake)
         session.flush()
+        session.expunge_all()
```
Comment (adrien-berchet) [inline tests/test_functional.py]: @elemoine , @zzzeek I had to add several `session.expunge_all()` calls to make the tests pass in PyPy... but I don't understand why it is required now while it was not before. Do you have any clue?

**Comment by adrien-berchet** [comment]
I tried to refactor the Raster and RasterElement classes. The Raster is now very close to the Geometry type and elements.

File: geoalchemy2/functions.py [hunk 2]
```diff
@@ -92,7 +93,20 @@ def __init__(self, *args, **kwargs):
...
+                if elem.extended:
+                    func_name = elem.geom_from_extended_version
+                    func_args = [elem.data]
+                else:
+                    func_name = elem.geom_from
+                    func_args = [elem.data, elem.srid]
+                new_arg = getattr(functions.func, func_name)(*func_args)
+            else:
+                new_arg = elem
+            new_args.append(new_arg)
```
Comment (elemoine) [inline geoalchemy2/functions.py:108]: Or maybe a bit more elegantly:
```py
args = list(args)
if expr is not None:
    args = [expr] + args
for idx, elem in enumerate(args):
    if isinstance(elem, elements.HasFunction):
        if elem.extended:
            func_name = elem.geom_from_extented_version
            func_args = [elem.data]
        else:
            func_name = elem.geom_from
            func_args = [elem.data, elem.srid]
        args[idx] = getattr(functions.func, func_name)(*func_args)
functions.GenericFunction.__init__(self, *args, **kwargs)
```

File: tests/test_functional.py (hunk 1 above)
Comment (elemoine) [reply tests/test_functional.py:361]: I have no clue unfortunately. I see nothing in your changes that could have an impact on that.

**Comment by elemoine** [comment]
Thanks a lot @adrien-berchet. That's great work! I have no idea about the need for `expunge_all` calls with PyPy. That's rather strange.

File: geoalchemy2/elements.py [hunk 3]
```diff
@@ -81,39 +85,19 @@ def __getstate__(self):
             'srid': self.srid,
             'data': str(self),
             'extended': self.extended,
-            'name': self.name,
         }
         return state
 
     def __setstate__(self, state):
-        self.__dict__.update(state)
```
Comment (elemoine) [inline geoalchemy2/elements.py:89]: That might be the cause of the behavior change with PyPy. I am not fully sure, but that's my best intuition for the moment.

File: tests/test_functional.py (hunk 1 above)
Comment (elemoine) [reply tests/test_functional.py]: @adrien-berchet I've just added a commit that removes the call to `expunge_all`, in favor to calls to `expire`, which, I think makes more sense, and `expire` is what's used elsewhere in this test file.

File: geoalchemy2/functions.py (hunk 2 above)
Comment (elemoine) [reply geoalchemy2/functions.py]: I've added a commit for that.

**Comment by elemoine** [comment]
I think this is good to go @adrien-berchet. As for PyPy I think the calls to `expire` in `test_functional.py` make sense, although I am not sure why the behavior is different with CPython and PyPy.

File: geoalchemy2/functions.py (hunk 2 above)
Comment (adrien-berchet) [reply geoalchemy2/functions.py]: It is much better indeed, thanks :-)

**Comment by adrien-berchet** [comment]
Thanks @elemoine !
I added some comments so I will merge when the CI is complete.

File: geoalchemy2/functions.py (hunk 2 above)
Comment (elemoine) [reply geoalchemy2/functions.py]: I'd say "a bit nicer" :-)

**Pull request merged**